### PR TITLE
Adding ServiceConnectionPolicies resource in NetworkConnectivity.

### PR DESCRIPTION
**Pull request by diogoEsteves42**
Adding two new resources in TPG.
Service Connection Policy requires a Service Class
https://cloud.google.com/vpc/docs/about-service-connection-policies

We can find information about the resources in the explorer file
https://cloud.google.com/network-connectivity/docs/reference/networkconnectivity/rest

This PR solves: https://github.com/hashicorp/terraform-provider-google/issues/15051

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Generated Terraform providers](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/generate-providers.md), and ran [`make test` and `make lint`](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/run-provider-tests/#run-unit-tests) in the generated providers to ensure it passes unit and linter tests.
- [x] [Ran](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/run-provider-tests.md) relevant acceptance tests using my own Google Cloud project and credentials (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read [Write release notes](https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
google_network_connectivity_service_connection_policy
```

### CODE REVIEW

**Comment by diogoEsteves42** [comment]
Hi, @roaks3  I have one question about this PR
I see that the product `networkConnectivity` was started via DCL, now when I create new resources using MMv1,
it looks like I have to remove the reference from the DCL in the file `mmv1/third_party/terraform/fwmodels/provider_model.go.erb`, is that approach correct?

It's a draft, I need to fix the test scenario for one resource.

**Comment by diogoEsteves42** [comment]
Hi @roaks3, just opened the PR to review. But I need some help with the serviceClass resource.

**Comment by roaks3** [comment]
Hi @diogoEsteves42 , this repo (used for GA and beta providers) requires that APIs be public. If `serviceClass` is not available in a public preview until next year, we would need to wait before making that addition here.

Could you confirm that the API you need is not currently public? Any other information you can provide would also help (like if you have engaged with a TAM).

**Comment by diogoEsteves42** [comment]
Hi @roaks3 , thanks for the answer. The last information that I have about `serviceClass` is that it will be available to users without allowlisting on early next year.

Can I kind of mock this resource? Use it with a static value? Because the `serviceConnectionPolicy` it's ready, but I don't know how to handle the dependency as it's required field. I'll update the PR by removing the `serviceClass` but I need to know how to make the test pass.

appreciate any help

File: mmv1/third_party/terraform/tests/resource_network_connectivity_service_connection_policies_test.go
Comment (diogoEsteves42) [inline mmv1/third_party/terraform/tests/resource_network_connectivity_service_connection_policies_test.go:83]: @roaks3 I talked with someone from the team of PSC and they told me that I can use this hardcode value to make the tests pass.

**Comment by diogoEsteves42** [comment]
Hi @roaks3 I've just made a simple update in the code, but I haven't removed the service class yet.

One question:
If the resource is published in the discovery file for v1, is this not enough to create a product here?

If it's not possible I'll update the PR and then we can follow the review.
Appreciate the help

**Comment by roaks3** [comment]
Hey @diogoEsteves42 , since the API is currently marked as Pre-GA, and uses an allowlist, it would not be something we can add to either of the public providers. However, I can follow up with the team internally to verify that those assumptions are correct.

**Comment by diogoEsteves42** [comment]
Hi @roaks3 another person is facing the same issue with some feature or resource not being public and he mentions something about releasing it in another way.

I will remove the serviceClass and update the tests, I'll @ you when it's ready.

**Comment by diogoEsteves42** [comment]
@roaks3 updated the PR, ready to review.

File: mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml
Comment (roaks3) [inline mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml:19]: Can be removed, this is the default

File: mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml
Comment (roaks3) [inline mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml:53]: Same comment, can be removed

File: mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml
Comment (roaks3) [inline mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml:75]: Some of these properties seem required, like network. I'm curious about the behavior of `location` as well, if it is not required.

**Review by roaks3 — COMMENTED** [review]
I took a first pass and had a few comments on the yaml. Will wait to see what comes back from the tests before continuing review. Thanks for the contribution!

File: mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml
Comment (diogoEsteves42) [reply mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml]: removed, thanks.

File: mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml
Comment (diogoEsteves42) [reply mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml]: removed, thanks.

File: mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml
Comment (diogoEsteves42) [reply mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml:74]: Makes sense, my bad, added the required for that.

File: mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml
Comment (roaks3) [inline mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml:86]: Per my previous comment, is `serviceClass` not required? It seemed like you needed to specify one when you were working out the tests.

File: mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml
Comment (roaks3) [inline mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml:95]: Same here, is it possible to create this resource without specifying a network?

File: mmv1/products/networkconnectivity/product.yaml
Comment (roaks3) [inline mmv1/products/networkconnectivity/product.yaml:1]: Nit: Same, can be bumped to 2023

File: mmv1/templates/terraform/examples/network_connectivity_policy_basic.tf.erb
Comment (roaks3) [inline mmv1/templates/terraform/examples/network_connectivity_policy_basic.tf.erb:28]: Can you briefly explain how this value is working? This will show up in the official documentation for this resource, and it will be used for testing in a number of different environments, so I'd like to understand if `gcp-memorystore-redis` is intended to be used in all of those places.

File: mmv1/third_party/terraform/fwmodels/provider_model.go.erb
Comment (roaks3) [inline mmv1/third_party/terraform/fwmodels/provider_model.go.erb:56]: Reminder for myself to follow up on the DCL piece you mentioned. From looking at the generated diffs, I think what you've done is fine, but want to make sure.

File: mmv1/third_party/terraform/tests/resource_network_connectivity_service_connection_policies_test.go
Comment (roaks3) [inline mmv1/third_party/terraform/tests/resource_network_connectivity_service_connection_policies_test.go:53]: Nit: If you have the time and feel like it, we have a pattern that works a little better when there are more arguments, where you use a `context` variable and `NPrintf()`. It mainly helps prevent mistakes if the positional arguments are passed in the wrong order. However, it's definitely ok to leave it as-is.

File: mmv1/third_party/terraform/tests/resource_network_connectivity_service_connection_policies_test.go
Comment (roaks3) [inline mmv1/third_party/terraform/tests/resource_network_connectivity_service_connection_policies_test.go:82]: Nit: It usually covers a few more edge cases to have a test like this specify the minimum set of fields (so you could probably remove `description`, maybe others). It will also help validate some of my early questions regarding which fields are actually required.

**Review by roaks3 — COMMENTED** [review]
Left a few more comments, I'll check in on the DCL piece. Otherwise things look good and 👍 tests are passing

File: mmv1/templates/terraform/examples/network_connectivity_policy_basic.tf.erb
Comment (diogoEsteves42) [reply mmv1/templates/terraform/examples/network_connectivity_policy_basic.tf.erb:28]: Ohh, great, I just forgot that.
Yeah this should not be public. I've talked with a person for the serviceConnectionPolicy and mentioned that I need a way to make the tests work. So he gave me this resource to use. So, it's not for all those places, just here, temporarily while we don't support service_class.

How can I set this one as an env var?

File: mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml
Comment (diogoEsteves42) [reply mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml:95]: No it's not, thanks for that.

File: mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml
Comment (diogoEsteves42) [reply mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml]: done!

File: mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml
Comment (diogoEsteves42) [reply mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml:86]: Done!

File: mmv1/third_party/terraform/tests/resource_network_connectivity_service_connection_policies_test.go
Comment (diogoEsteves42) [reply mmv1/third_party/terraform/tests/resource_network_connectivity_service_connection_policies_test.go]: Do you believe this test is not necessary to exist? I'm just adding labels to test the  update.

File: mmv1/third_party/terraform/fwmodels/provider_model.go.erb
Comment (roaks3) [reply mmv1/third_party/terraform/fwmodels/provider_model.go.erb:56]: I think this is technically ok, but it would be safer to add a `tpgtools_product.yaml` file to `tpgtools/overrides/networkconnectivity` that looks like [this](https://github.com/GoogleCloudPlatform/magic-modules/pull/6608/files#diff-de1e23c69df2f622c191e8c6a721f08bbcbdda0cdb3bd1d1b4c4343e9e648816). Could you try that out, in addition to this line? I think it will clean up some other code that is now redundant.

File: mmv1/third_party/terraform/tests/resource_network_connectivity_service_connection_policies_test.go
Comment (roaks3) [reply mmv1/third_party/terraform/tests/resource_network_connectivity_service_connection_policies_test.go]: We actually have a requirement to test updating any fields that are updatable, so I would definitely keep this, but my suggestion is to simply remove the `description` on this line so that the test hits a few more edges of the implementation. You may also be able to remove other fields here that are not marked required, but I'm not sure.

**Comment by roaks3** [comment]
Your responses look good, and I made 2 more comments above. After the next set of changes, I'll kick off the tests, and we should be very close to being able to merge this.

File: mmv1/third_party/terraform/fwmodels/provider_model.go.erb
Comment (diogoEsteves42) [reply mmv1/third_party/terraform/fwmodels/provider_model.go.erb:56]: @roaks3 , great, added the file that you mention.

**Comment by diogoEsteves42** [comment]
@roaks3, great thanks for the help with this resource.

I've added the updates, just need to validate about the env_var for the service_class field.

File: mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml
Comment (roaks3) [inline mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml:69]: I think you may need `immutable: true` on `location`, `serviceClass`, `network`, and `pscConfig.subnetworks`. Possibly `pscConfig.limit` too. Unfortunately, the discovery docs do not always describe all of these constraints, so you'd have to try changing them and see what happens.

This ties in to the `_update` test you have below: if a field can be updated through the API, the update test should test that by modifying its value. So in other words, if any of the fields I mentioned above ARE NOT immutable, we should see their value being changed in the update test.

File: mmv1/third_party/terraform/tests/resource_network_connectivity_service_connection_policies_test.go
Comment (roaks3) [inline mmv1/third_party/terraform/tests/resource_network_connectivity_service_connection_policies_test.go:14]: Nit: spacing got a little messed up here

**Review by roaks3 — CHANGES_REQUESTED** [review]

File: mmv1/templates/terraform/examples/network_connectivity_policy_basic.tf.erb
Comment (roaks3) [reply mmv1/templates/terraform/examples/network_connectivity_policy_basic.tf.erb]: Sorry, I missed this comment on the prior pass.

Could you share any more about this value? Is it the name of an actual resource, or is it some sort of fake value? If it is an actual resource, is it publicly accessible (for anyone who chooses to run these tests)? If any sort of access is needed for it, we will need someone to make sure that access is set up for all of our internal environments.

Comment (diogoEsteves42) [reply mmv1/templates/terraform/examples/network_connectivity_policy_basic.tf.erb]: @roaks3 according to someone from the ServiceConnectionPolicies, this is a real resource, it's not a fake one, that it will be used while we don't have the serviceClass public available. It's not publicly accessible and I don't know how to make sure that all the envs will be added to have access. I will ask him to talk with you about that.

Comment (roaks3) [reply mmv1/templates/terraform/examples/network_connectivity_policy_basic.tf.erb]: Ok, sounds good. Depending on that conversation, we may want to hold this from being merged, or skip the tests right before we merge, but in the meantime:

- We will definitely want the docs to have some sort of example value. You can do this by adding a new field under `vars` that specifies the service class value that we would use in the docs (like `example-class`). Then you can also add a `test_vars_overrides` block with the value that will be used for tests only (you can search for some existing examples of this).
- We tend to avoid environment variables when possible, because we need to manage them centrally, and it would become overwhelming if they were needed by many teams/resources. Normally in cases like this, we would keep the test value as you have it (a hardcoded value), unless there is some sort of concern around doing that.

File: mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml
Comment (diogoEsteves42) [reply mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml:72]: Oh got, my bad, I was just paying attention to the required field.

File: mmv1/third_party/terraform/tests/resource_network_connectivity_service_connection_policies_test.go
Comment (diogoEsteves42) [reply mmv1/third_party/terraform/tests/resource_network_connectivity_service_connection_policies_test.go]: Great, fixed.

File: mmv1/templates/terraform/examples/network_connectivity_policy_basic.tf.erb
Comment (diogoEsteves42) [reply mmv1/templates/terraform/examples/network_connectivity_policy_basic.tf.erb]: Ohh, I got the idea.

I've added some more code to help with env_var.

File: mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml
Comment (diogoEsteves42) [reply mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml:73]: Hi @roaks3, I've managed to validate all the fields if they are immutable or not. Tests are passing again.

File: mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml
Comment (roaks3) [inline mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml:63]: I spoke with the engineer internally, and this can be changed to `service_class_name: 'gcp-memorystore-redis'`. The hardcoded value is fine for all test environments, and so you can also remove the other related code for the environment variable.

File: mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml
Comment (roaks3) [reply mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml:102]: The other fields look good, but I believe `network` here still needs to be marked as `immutable`.

File: mmv1/templates/terraform/examples/network_connectivity_policy_basic.tf.erb
Comment (roaks3) [inline mmv1/templates/terraform/examples/network_connectivity_policy_basic.tf.erb:6]: Just to verify, I don't see the consumer network/subnetwork used elsewhere in this test (same with the update test). Is that intended?

Comment (diogoEsteves42) [reply mmv1/templates/terraform/examples/network_connectivity_policy_basic.tf.erb:6]: That's right, I was just adding that because this resource is part of one product that brings this concept of connecting two different VPCs, a consumer side and a producer side. And it's a basic scenario, I can remove that.
I just keep that in here, because we can grow this basic example to a more complex one.

File: mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml
Comment (diogoEsteves42) [reply mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml:102]: @roaks3 I was testing this scenario, if I add the network as immutable, the test fails because the API requires the network in the payload.

File: mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml
Comment (diogoEsteves42) [reply mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml:63]: Great, done that.

Comment (roaks3) [reply mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml]: Would you mind changing just this section back to still use a different value for docs? For example:
```
      service_class_name: 'my-basic-service-class'
    test_vars_overrides:
      service_class_name: 'gcp-memorystore-redis'
```

Comment (diogoEsteves42) [reply mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml]: Of course, it makes sense to keep this way.

File: mmv1/templates/terraform/examples/network_connectivity_policy_basic.tf.erb
Comment (diogoEsteves42) [reply mmv1/templates/terraform/examples/network_connectivity_policy_basic.tf.erb]: yeah, makes sense. Done that.

**Comment by roaks3** [comment]
Check failure:
```
gofmt needs running on the following files:
./google-beta/resource_network_connectivity_service_connection_policies_test.go
make: *** [GNUmakefile:23: fmtcheck] Error 1
You can use the command: `make fmt` to reformat code.
```

Yaml validation failures:
```
repo/mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml
  Error: 110:16 [trailing-spaces] trailing spaces

repo/mmv1/products/networkconnectivity/product.yaml
  Error: 2[7](https://github.com/GoogleCloudPlatform/magic-modules/actions/runs/5718825289/job/15495451001#step:6:8):5 [new-line-at-end-of-file] no new line character at the end of file
  Error: 27:1 [trailing-spaces] trailing spaces
```

I'm not used to seeing these types of error fail the whole job, but we may need to fix the above before the tests will run.

**Comment by diogoEsteves42** [comment]
@roaks3 I believe it was the value for the variable override for the tests. Fixed, ran the linter and both tests are okay.

File: mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml
Comment (roaks3) [reply mmv1/products/networkconnectivity/ServiceConnectionPolicies.yaml:99]: Hmm, that's strange. Could you provide any more info on that? I think we should be able to configure this property in a way that works.

Alternatively, you could submit it with the `immutable: true` and then I can see the failure in the logs in VCR.

File: mmv1/third_party/terraform/tests/resource_network_connectivity_service_connection_policies_test.go
Comment (roaks3) [inline mmv1/third_party/terraform/tests/resource_network_connectivity_service_connection_policies_test.go:68]: Nit: You should be able to remove this subnet because it isn't being used in this config. I don't think it will be faster, but a little cleaner to read.

File: mmv1/third_party/terraform/tests/resource_network_connectivity_service_connection_policies_test.go
Comment (roaks3) [inline mmv1/third_party/terraform/tests/resource_network_connectivity_service_connection_policies_test.go:95]: Nit: Similarly, you should be able to remove this subnet because it isn't being used in this config

**Review by roaks3 — COMMENTED** [review]
Looks good! I would just like to make sure the `network` field is in a good place so that we don't need to make changes later.

File: mmv1/third_party/terraform/tests/resource_network_connectivity_service_connection_policies_test.go
Comment (diogoEsteves42) [reply mmv1/third_party/terraform/tests/resource_network_connectivity_service_connection_policies_test.go]: great, removed.

File: mmv1/third_party/terraform/tests/resource_network_connectivity_service_connection_policies_test.go
Comment (diogoEsteves42) [reply mmv1/third_party/terraform/tests/resource_network_connectivity_service_connection_policies_test.go]: got it, removed.

**Comment by roaks3** [comment]
Looking at the error, it does appear that `network` is required on the PATCH call, but is not being sent when it is `immutable`. The best way to handle this would be to add an encoder similar to [encoders/apigee_organization.go.erb](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/templates/terraform/encoders/apigee_organization.go.erb), which would explicitly set the network on the request obj before being sent in the PATCH. You can configure the encoder with a `custom_code` block in the yaml, and a `update_encoder` entry under that.

I think this will fix the issue with updates, while allowing `network` to remain immutable. Let me know how it goes.

**Comment by diogoEsteves42** [comment]
Owww, that works, really glad that you helped me @roaks3 
Amazing, I believe that now everything is set as expected by the API and will be fine on Terraform.

Thanks again, without your help I would spend much more time on this.

**Review by roaks3 — APPROVED** [review]
LGTM, thanks for working through all the feedback.

I also tweaked the release note to remove the service class resource.

**Pull request merged**
